server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java

```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
         bulkProcessor.close();
     }
 
+    public void testConcurrentExecutions() throws Exception {
+        final AtomicBoolean called = new AtomicBoolean(false);
+        int estimatedTimeForTest = Integer.MAX_VALUE;
+        final int simulateWorkTimeInMillis = 5;
+        int concurrentClients= 0;
+        int concurrentBulkRequests = 0;
+        int expectedExecutions= 0;
+        int maxBatchSize = 0;
+        int maxDocuments = 0;
+        int iterations = 0;
+        boolean runTest = true;
+        //find some randoms that allow this test to take under ~ 10 seconds
+        while (estimatedTimeForTest > 10_000) {
+            if (iterations++ > 1_000) { //extremely unlikely
+                runTest = false;
+            }
+            maxBatchSize = randomIntBetween(1, 100);
+            maxDocuments = randomIntBetween(maxBatchSize, 1_000_000);
+            concurrentClients = randomIntBetween(1, 20);
+            concurrentBulkRequests = randomIntBetween(0, 20);
+            expectedExecutions = maxDocuments / maxBatchSize;
+            estimatedTimeForTest = (expectedExecutions * simulateWorkTimeInMillis) /
+                Math.min(concurrentBulkRequests + 1, concurrentClients);
+        }
+        assumeTrue("failed to find random values that allows test to run quickly", runTest);
+        BulkResponse bulkResponse = new BulkResponse(new BulkItemResponse[]{new BulkItemResponse()}, 0);
+        AtomicInteger failureCount = new AtomicInteger(0);
+        AtomicInteger successCount = new AtomicInteger(0);
+        AtomicInteger requestCount = new AtomicInteger(0);
+        AtomicInteger docCount = new AtomicInteger(0);
+        BiConsumer<BulkRequest, ActionListener<BulkResponse>> consumer = (request, listener) ->
+        {
+            try {
+                Thread.sleep(simulateWorkTimeInMillis); //simulate work
+            } catch (InterruptedException e) {
+                failureCount.getAndIncrement();
+                logger.error("interrupted while sleeping. There is likely something wrong with this test!", e);
+            }
+            listener.onResponse(bulkResponse);
+        };
+        BulkProcessor bulkProcessor = new BulkProcessor(consumer, BackoffPolicy.noBackoff(),
```

Comment:
I think it is worth replacing the explicit close of the bulk processor below with a try-with-resources

```suggestion
        try (BulkProcessor bulkProcessor = new BulkProcessor(consumer, BackoffPolicy.noBackoff(),
```